tests/test_tiatoolbox.py

```diff
@@ -123,37 +152,621 @@ def test_wsireader_slide_info(_response_svs, tmp_path):
+
+def test__relative_level_scales_jp2_level(_response_jp2):
+    """Test jp2 calculation of relative level scales for level."""
+    wsi = wsireader.OmnyxJP2WSIReader(_response_jp2)
+    level_scales = wsi._relative_level_scales(3, "level")
+    level_scales = np.array(level_scales)
+    assert np.array_equal(level_scales[3], [1, 1])
+    downsamples = np.array(wsi.slide_info.level_downsamples)
+    expected = downsamples / downsamples[3]
+    assert np.array_equal(level_scales[:, 0], level_scales[:, 1])
+    assert np.array_equal(level_scales[:, 0], expected)
+
+
+def test__relative_level_scales_openslide_level_float(_response_ndpi):
+    """Test openslide calculation of relative level scales for fracitonal level."""
+    wsi = wsireader.OpenSlideWSIReader(_response_ndpi)
+    level_scales = wsi._relative_level_scales(1.5, "level")
+    level_scales = np.array(level_scales)
+    assert level_scales[0] == approx([1 / 3, 1 / 3])
+    downsamples = np.array(wsi.slide_info.level_downsamples)
+    expected = downsamples / downsamples[0] * (1 / 3)
+    assert np.array_equal(level_scales[:, 0], level_scales[:, 1])
+    assert np.array_equal(level_scales[:, 0], expected)
+
+
+def test__relative_level_scales_jp2_level_float(_response_jp2):
+    """Test jp2 calculation of relative level scales for fracitonal level."""
+    wsi = wsireader.OmnyxJP2WSIReader(_response_jp2)
+    level_scales = wsi._relative_level_scales(1.5, "level")
+    level_scales = np.array(level_scales)
+    assert level_scales[0] == approx([1 / 3, 1 / 3])
+    downsamples = np.array(wsi.slide_info.level_downsamples)
+    expected = downsamples / downsamples[0] * (1 / 3)
+    assert np.array_equal(level_scales[:, 0], level_scales[:, 1])
+    assert np.array_equal(level_scales[:, 0], expected)
+
+
+def test__relative_level_scales_invalid_units(_response_svs):
+    """Test _relative_level_scales with invalid units."""
+    wsi = wsireader.OpenSlideWSIReader(_response_svs)
+    with pytest.raises(ValueError):
+        wsi._relative_level_scales(1.0, "gibberish")
+
+
+def test__relative_level_scales_no_mpp():
+    """Test _relative_level_scales objective when mpp is None."""
+
+    class DummyWSI:
+        _relative_level_scales = wsireader.WSIReader._relative_level_scales
+
+        @property
+        def slide_info(self):
+            return wsireader.WSIMeta((100, 100))
+
+    wsi = DummyWSI()
+    with pytest.raises(ValueError):
+        wsi._relative_level_scales(1.0, "mpp")
+
+
+def test__relative_level_scales_no_objective_power():
+    """Test _relative_level_scales objective when objective power is None."""
+
+    class DummyWSI:
+        _relative_level_scales = wsireader.WSIReader._relative_level_scales
+
+        @property
+        def slide_info(self):
+            return wsireader.WSIMeta((100, 100))
+
+    wsi = DummyWSI()
+    with pytest.raises(ValueError):
+        wsi._relative_level_scales(10, "power")
+
+
+def test__relative_level_scales_level_too_high(_response_svs):
```

Comment:
why there are two underscores after `test`? If we need to keep two underscores then we should keep it consistent.

Reply:
It's because the function being tested is a private function, `_relative_level_scales`. This was changed by automatic refactoring. Can remove double underscore from test although it does indicate that it is private.

Reply:
it's fine then leave it.